// react-native-wgpu/lib/commonjs/hooks.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.warnIfNotHardwareAccelerated = exports.useSurface = exports.useMainDevice = exports.useDevice = exports.useCanvasRef = exports.GPUDeviceProvider = void 0;
var _react = _interopRequireWildcard(require("react"));
function _interopRequireWildcard(e, t) { if ("function" == typeof WeakMap) var r = new WeakMap(), n = new WeakMap(); return (_interopRequireWildcard = function (e, t) { if (!t && e && e.__esModule) return e; var o, i, f = { __proto__: null, default: e }; if (null === e || "object" != typeof e && "function" != typeof e) return f; if (o = t ? n : r) { if (o.has(e)) return o.get(e); o.set(e, f); } for (const t in e) "default" !== t && {}.hasOwnProperty.call(e, t) && ((i = (o = Object.defineProperty) && Object.getOwnPropertyDescriptor(e, t)) && (i.get || i.set) ? o(f, t, i) : f[t] = e[t]); return f; })(e, t); }
const warnIfNotHardwareAccelerated = adapter => {
  // Check if adapter is a fallback adapter using the new GPUAdapterInfo API
  // Note: isFallbackAdapter was moved from GPUAdapter to GPUAdapterInfo in Chrome 140
  if (adapter.info && adapter.info.isFallbackAdapter) {
    console.warn("GPUAdapter is not hardware accelerated. This is common on Android emulators. Rendering will be slow. Some features may be unavailable.");
  }
};
exports.warnIfNotHardwareAccelerated = warnIfNotHardwareAccelerated;
const DeviceContext = /*#__PURE__*/(0, _react.createContext)(null);
const GPUDeviceProvider = ({
  children,
  adapterOptions,
  deviceDescriptor
}) => {
  const state = useDevice(adapterOptions, deviceDescriptor);
  if (!state.device) {
    return null;
  }
  return /*#__PURE__*/_react.default.createElement(DeviceContext.Provider, {
    value: state
  }, children);
};
exports.GPUDeviceProvider = GPUDeviceProvider;
const useSurface = () => {
  const ref = (0, _react.useRef)(null);
  const [surface, setSurface] = (0, _react.useState)(null);
  (0, _react.useLayoutEffect)(() => {
    const sur = ref.current.getNativeSurface();
    setSurface(sur);
  }, []);
  return {
    ref,
    surface
  };
};
exports.useSurface = useSurface;
const useMainDevice = () => {
  const ctx = (0, _react.useContext)(DeviceContext);
  if (!ctx) {
    throw new Error("No DeviceContext found.");
  }
  return ctx;
};
exports.useMainDevice = useMainDevice;
const useCanvasRef = () => (0, _react.useRef)(null);
exports.useCanvasRef = useCanvasRef;
const useDevice = (adapterOptions, deviceDescriptor) => {
  const [state, setState] = (0, _react.useState)(null);
  (0, _react.useEffect)(() => {
    (async () => {
      if (!state) {
        const adapter = await navigator.gpu.requestAdapter(adapterOptions);
        if (!adapter) {
          throw new Error("No appropriate GPUAdapter found.");
        }
        warnIfNotHardwareAccelerated(adapter);
        const device = await adapter.requestDevice(deviceDescriptor);
        if (!device) {
          throw new Error("No appropriate GPUDevice found.");
        }
        setState({
          adapter,
          device
        });
        return;
      }
    })();
  }, [adapterOptions, deviceDescriptor, state]);
  return {
    adapter: state?.adapter ?? null,
    device: state?.device ?? null
  };
};
exports.useDevice = useDevice;
