// react-native-wgpu/lib/module/hooks.js
import React, { createContext, useContext, useEffect, useLayoutEffect, useRef, useState } from "react";
export const warnIfNotHardwareAccelerated = adapter => {
  // Check if adapter is a fallback adapter using the new GPUAdapterInfo API
  // Note: isFallbackAdapter was moved from GPUAdapter to GPUAdapterInfo in Chrome 140
  if (adapter.info && adapter.info.isFallbackAdapter) {
    console.warn("GPUAdapter is not hardware accelerated. This is common on Android emulators. Rendering will be slow. Some features may be unavailable.");
  }
};
const DeviceContext = /*#__PURE__*/createContext(null);
export const GPUDeviceProvider = ({
  children,
  adapterOptions,
  deviceDescriptor
}) => {
  const state = useDevice(adapterOptions, deviceDescriptor);
  if (!state.device) {
    return null;
  }
  return /*#__PURE__*/React.createElement(DeviceContext.Provider, {
    value: state
  }, children);
};
export const useSurface = () => {
  const ref = useRef(null);
  const [surface, setSurface] = useState(null);
  useLayoutEffect(() => {
    const sur = ref.current.getNativeSurface();
    setSurface(sur);
  }, []);
  return {
    ref,
    surface
  };
};
export const useMainDevice = () => {
  const ctx = useContext(DeviceContext);
  if (!ctx) {
    throw new Error("No DeviceContext found.");
  }
  return ctx;
};
export const useCanvasRef = () => useRef(null);
export const useDevice = (adapterOptions, deviceDescriptor) => {
  const [state, setState] = useState(null);
  useEffect(() => {
    (async () => {
      if (!state) {
        const adapter = await navigator.gpu.requestAdapter(adapterOptions);
        if (!adapter) {
          throw new Error("No appropriate GPUAdapter found.");
        }
        warnIfNotHardwareAccelerated(adapter);
        const device = await adapter.requestDevice(deviceDescriptor);
        if (!device) {
          throw new Error("No appropriate GPUDevice found.");
        }
        setState({
          adapter,
          device
        });
        return;
      }
    })();
  }, [adapterOptions, deviceDescriptor, state]);
  return {
    adapter: state?.adapter ?? null,
    device: state?.device ?? null
  };
};
